### PR TITLE
Remove more old backend ISA concepts

### PR DESCRIPTION
**Pull request by bnjbvr**
This also paves the way for unifying `TargetIsa` and `MachBackend`, since now they map one to one. In theory the two traits could be merged, which would be nice to limit the number of total concepts. Also they have quite different responsibilities, so it might be fine to keep them separate. Any thoughts @cfallin?

Interestingly, this PR started as removing`RegInfo` from the `TargetIsa` trait since the adapter returned a dummy value there. From the fallout, noticed that all `Display` implementations didn't needed an ISA anymore (since these were only used to render ISA specific registers). Also the whole family of `RegInfo` / `ValueLoc` / `RegUnit` was exclusively used for the old backend, and these could be removed. Notably, some IR instructions needed to be removed, because they were using `RegUnit` too: this was the oddball of `regfill` / `regmove` / `regspill` / `copy_special`, which were IR instructions inserted by the old regalloc. Fare thee well!

(feel free to squash the commits before merging, the history of this PR is a bit messy right now)

### CODE REVIEW

**Review by cfallin — APPROVED** [review]
All of this looks great, thanks!

> `TargetIsa` vs `MachBackend`

Yes, I agree that we should merge these eventually; there's no reason to have the adapter anymore. It might make sense just to rename `MachBackend` to `TargetIsa`, after moving over whatever remaining methods it needs, so that external API users that refer to a `TargetIsa` don't have to change.

File: cranelift/codegen/src/ir/extfunc.rs [hunk 1]
```diff
@@ -51,8 +51,8 @@ impl Signature {
     }
 
     /// Return an object that can display `self` with correct register names.
-    pub fn display<'a, R: Into<Option<&'a RegInfo>>>(&'a self, regs: R) -> DisplaySignature<'a> {
-        DisplaySignature(self, regs.into())
+    pub fn display(&self) -> DisplaySignature<'_> {
```
Comment (bjorn3) [inline cranelift/codegen/src/ir/extfunc.rs:54]: Can you directly implement Display for Signature?

File: cranelift/codegen/src/ir/extfunc.rs [hunk 2]
```diff
@@ -217,13 +201,13 @@ impl AbiParam {
     }
 
     /// Return an object that can display `self` with correct register names.
-    pub fn display<'a, R: Into<Option<&'a RegInfo>>>(&'a self, regs: R) -> DisplayAbiParam<'a> {
-        DisplayAbiParam(self, regs.into())
+    pub fn display<'a>(&'a self) -> DisplayAbiParam<'a> {
```
Comment (bjorn3) [inline cranelift/codegen/src/ir/extfunc.rs:204]: Same

File: cranelift/codegen/src/isa/mod.rs [hunk 3]
```diff
@@ -268,66 +299,35 @@ pub trait TargetIsa: fmt::Display + Send + Sync {
...
-
-    /// IntCC condition for Unsigned Addition Overflow (Carry).
-    fn unsigned_add_overflow_condition(&self) -> ir::condcodes::IntCC;
-
-    /// IntCC condition for Unsigned Subtraction Overflow (Borrow/Carry).
-    fn unsigned_sub_overflow_condition(&self) -> ir::condcodes::IntCC;
-
     /// Returns the flavor of unwind information emitted for this target.
-    fn unwind_info_kind(&self) -> UnwindInfoKind {
+    pub(crate) fn unwind_info_kind(&self) -> UnwindInfoKind {
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/mod.rs:330]: I think I use this in cg_clif.

File: cranelift/frontend/src/frontend.rs [hunk 4]
```diff
@@ -585,8 +585,8 @@ impl<'a> FunctionBuilder<'a> {
     /// Useful for debug purposes. Use it with `None` for standard printing.
     // Clippy thinks the lifetime that follows is needless, but rustc needs it
     #[cfg_attr(feature = "cargo-clippy", allow(clippy::needless_lifetimes))]
-    pub fn display<'b, I: Into<Option<&'b dyn TargetIsa>>>(&'b self, isa: I) -> DisplayFunction {
-        self.func.display(isa)
+    pub fn display<'b>(&'b self) -> DisplayFunction {
```
Comment (bjorn3) [inline cranelift/frontend/src/frontend.rs:588]: Same re Display.

File: cranelift/codegen/src/ir/extfunc.rs (hunk 1 above)
Comment (bnjbvr) [reply cranelift/codegen/src/ir/extfunc.rs:54]: Indeed, thanks!

File: cranelift/codegen/src/isa/mod.rs (hunk 3 above)
Comment (bnjbvr) [reply cranelift/codegen/src/isa/mod.rs:330]: Kept as i.

**Pull request merged**